aws/credentials/process_provider.go

```diff
@@ -0,0 +1,228 @@
+	Filename string
+
+	// AWS Profile to retrieve the 'credential_process' from. If empty
+	// will default to environment variable "AWS_PROFILE" or "default" if
+	// environment variable is also not set.
+	Profile string
+
+	// ExpiryWindow will allow the credentials to trigger refreshing prior to
+	// the credentials actually expiring. This is beneficial so race conditions
+	// with expiring credentials do not cause request to fail unexpectedly
+	// due to ExpiredTokenException exceptions.
+	//
+	// So a ExpiryWindow of 10s would cause calls to IsExpired() to return true
+	// 10 seconds before the credentials are actually expired.
+	//
+	// If ExpiryWindow is 0 or less it will be ignored.
+	ExpiryWindow time.Duration
+
+	// A function to perform the credential_process execution. Can be replaced
+	// for testing
+	executionFunc func(process string) ([]byte, error)
+}
+
+// NewProcessProvider returns a pointer to a new Credentials object
+// wrapping the Profile file provider.
+func NewProcessProvider(filename, profile string) *Credentials {
+	return NewCredentials(&ProcessProvider{
+		Filename:      filename,
+		Profile:       profile,
+		executionFunc: executeCredentialProcess,
+	})
+}
+
+// Retrieve executes the 'credential_process' and returns the credentials
+func (p *ProcessProvider) Retrieve() (Value, error) {
+	filename, err := p.filename()
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, err
+	}
+
+	creds, err := p.loadProfile(filename, p.profile())
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, err
+	}
+
+	return creds, nil
+}
+
+// loadProfiles executes the 'credential_process' from the profile of the given
+// file, and returns the credentials retrieved. An error will be returned if it
+// fails to read from the file, or the data is invalid.
+func (p *ProcessProvider) loadProfile(filename, profile string) (Value, error) {
+	config, err := ini.Load(filename)
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "failed to load shared config file", err)
+	}
+
+	var iniProfile *ini.Section
+	if len(profile) == 0 || profile == "default" {
+		iniProfile, err = config.GetSection("default")
+	} else {
+		iniProfile, err = config.GetSection("profile " + profile)
+	}
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "failed to get profile", err)
+	}
+
+	process, err := iniProfile.GetKey("credential_process")
+	if err != nil {
+		return Value{ProviderName: ProcessProviderName}, awserr.New("ProcessProviderLoad", "No credential_process specified", err)
+	}
+	return p.credentialProcess(process.String())
+}
+
+type credentialProcessResponse struct {
+	Version         int
+	AccessKeyID     string `json:"AccessKeyId"`
+	SecretAccessKey string
+	SessionToken    string
+	Expiration      string
+}
+
+func executeCredentialProcess(process string) ([]byte, error) {
+	processArgs := strings.Split(process, " ")
```

Comment:
I think this split will have issue with quoted strings within the argument list of the `credential_process`.  I don't think the following will be handled correctly:

```
credential_process=ls "my cool/path with/spaces"
```

will be split into:
```go
[]string{"ls", "\"my", "cool/path", "with/spaces\""}
```